### PR TITLE
[ICD] Add Nonce validation to Check-In protocol and parsing unit tests

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Description
PR adds the nonce validation when we are parsing a Check-In message. We recalculate the nonce with the key and the received check-in counter to make sure the received nonce is the correct one.

PR adds all the necessary unit tests to test the Check-In protocol parsing.

#### Tests
Unit tests to validate implementation

### CODE REVIEW

File: src/protocols/secure_channel/tests/TestCheckinMsg.cpp [hunk 1]
```diff
@@ -203,16 +283,16 @@ void TestCheckInMsg::TestCheckInMessageGenerate_EmptyAesKeyHandle(nlTestSuite *
     ByteSpan applicationData(vector.application_data, vector.application_data_len);
 
     /*
-        Passing an empty key handle while using PSA crypto will result in a failure.
-        When using OpenSSL this same test result in a success.
-        Issue #28986
-
-    Verify that the generation fails with an empty key handle
+        TODO(#28986): Passing an empty key handle while using PSA crypto will result in a failure.
```
Comment (lpbeliveau-silabs) [inline src/protocols/secure_channel/tests/TestCheckinMsg.cpp:286]: Which one is the desired outcome?

File: src/protocols/secure_channel/CheckinMessage.h [hunk 2]
```diff
@@ -74,13 +74,16 @@ class DLL_EXPORT CheckinMessage
...
      * @param[in,out]   appData           The optional application data decrypted. The input size of appData must be at least the
-     * size of GetAppDataSize(payload) + sizeof(CounterType), because appData is used as a work buffer for the decryption process.
-     * The output size on success will be GetAppDataSize(payload).
+     *                                    size of GetAppDataSize(payload) + sizeof(CounterType), because appData is used as a work
+     * buffer for the decryption process. The output size on success will be GetAppDataSize(payload).
+     * If an error occurs, appData might countain data, but the data CANNOT be used since we were not able to validate it.
      *
      * @return CHIP_ERROR_INVALID_MESSAGE_LENGTH if the payload is shorter than the minimum payload size
      *         CHIP_ERROR_BUFFER_TOO_SMALL if appData buffer is too small
      *         CHIP_ERROR_INVALID_ARGUMENT if the provided arguments cannot be used to parse the Check-In message
```
Comment (lpbeliveau-silabs) [inline src/protocols/secure_channel/CheckinMessage.h:85]: It seems that the test is not currently testing for CHIP_ERROR_INVALID_ARGUMENT, can this error still be generated?

**Review by lpbeliveau — APPROVED** [review]

File: src/protocols/secure_channel/tests/TestCheckinMsg.cpp (hunk 1 above)
Comment (mkardous-silabs) [reply src/protocols/secure_channel/tests/TestCheckinMsg.cpp:286]: Most the likely the one where the encryption function returns an error but the answer is not definitive. 

As 0 key is a valid key but and undefined psa_key_id is not... The issue tracks the investigation and resolution but it is not clear cut.

File: src/protocols/secure_channel/CheckinMessage.h (hunk 2 above)
Comment (mkardous-silabs) [reply src/protocols/secure_channel/CheckinMessage.h]: removed since no longuer applicable

File: src/protocols/secure_channel/CheckinMessage.h [hunk 3]
```diff
@@ -74,13 +74,15 @@ class DLL_EXPORT CheckinMessage
      * @param[in]       hmac128KeyHandle  Key handle with which to verify the received nonce in the check-in payload (using HMAC).
      * @param[in]       payload           The received payload to decrypt and parse
      * @param[out]      counter           The counter value retrieved from the payload
+     *                                    If an error occurs, no value will be set.
      * @param[in,out]   appData           The optional application data decrypted. The input size of appData must be at least the
-     * size of GetAppDataSize(payload) + sizeof(CounterType), because appData is used as a work buffer for the decryption process.
-     * The output size on success will be GetAppDataSize(payload).
+     *                                    size of GetAppDataSize(payload) + sizeof(CounterType), because appData is used as a work
+     * buffer for the decryption process. The output size on success will be GetAppDataSize(payload).
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CheckinMessage.h:80]: The indent here is sort of odd...

**Review by bzbarsky-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

File: src/protocols/secure_channel/CheckinMessage.h (hunk 3 above)
Comment (mkardous-silabs) [reply src/protocols/secure_channel/CheckinMessage.h:80]: I agree but the restyler won't play nice and i have no idea how to work around this...

Comment (mkardous-silabs) [reply src/protocols/secure_channel/CheckinMessage.h]: Trailing white space was confusing the restyler. fixed the indenting

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**